Chinese_Version/ch_7_Finetune/7_1_Finetune_Llama2-7B.md

```diff
@@ -0,0 +1,295 @@
+
+### 7.1.2.4 加载Tokenizer
+
+分词器可以在 LLM 训练和推理中实现分词和去分词过程。您可以使用 [Huggingface Transformers](https://huggingface.co/docs/transformers/index) API来加载 LLM 推理需要的分词器，它可以与 BigDL-LLM 加载的模型无缝配合使用。对于Llama 2，对应的tokenizer类为`LlamaTokenizer`。
+
+```python
+from transformers import LlamaTokenizer
+tokenizer = LlamaTokenizer.from_pretrained(pretrained_model_name_or_path="meta-llama/Llama-2-7b-chat-hf", trust_remote_code=True)
+tokenizer.pad_token_id = 0
+tokenizer.padding_side = "left"
+```
+
+> **注意**
+>
+> 如果您已经下载了 Llama 2 (7B) 模型，您可以将 `pretrained_model_name_or_path` 指定为本地模型路径。
+
+### 7.1.2.5 进行训练
+
+现在您可以通过使用 HuggingFace 系统设置`trainer`来开始训练过程。 这里我们将`warmup_steps`设置为 20 以加速训练过程。
+```python
+import transformers
+trainer = transformers.Trainer(
+    model=model,
+    train_dataset=data["train"],
+    args=transformers.TrainingArguments(
+        per_device_train_batch_size=4,
+        gradient_accumulation_steps= 1,
+        warmup_steps=20,
+        max_steps=200,
+        learning_rate=2e-4,
+        save_steps=100,
+        fp16=True,
+        logging_steps=20,
+        output_dir="outputs", # 这里指定你自己的输出路径
+        optim="adamw_hf", # 尚不支持 paged_adamw_8bit优化
+        # gradient_checkpointing=True, # #可以进一步减少内存使用但速度较慢
+    ),
+    data_collator=transformers.DataCollatorForLanguageModeling(tokenizer, mlm=False),
+)
+model.config.use_cache = False  # 消除警告，进行推理时应重新启用
+result = trainer.train()
+```
+我们可以获得以下输出来展示我们的训练损失：
+```
+/home/arda/anaconda3/envs/yining-llm-qlora/lib/python3.9/site-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
+  warnings.warn(
+{'loss': 1.7193, 'learning_rate': 0.0002, 'epoch': 0.03}                                                             
+{'loss': 1.3242, 'learning_rate': 0.00017777777777777779, 'epoch': 0.06}                                             
+{'loss': 1.2266, 'learning_rate': 0.00015555555555555556, 'epoch': 0.1}                                              
+{'loss': 1.1534, 'learning_rate': 0.00013333333333333334, 'epoch': 0.13}                                             
+{'loss': 0.9368, 'learning_rate': 0.00011111111111111112, 'epoch': 0.16}                                             
+{'loss': 0.9321, 'learning_rate': 8.888888888888889e-05, 'epoch': 0.19}                                              
+{'loss': 0.9902, 'learning_rate': 6.666666666666667e-05, 'epoch': 0.22}                                              
+{'loss': 0.8593, 'learning_rate': 4.4444444444444447e-05, 'epoch': 0.26}                                             
+{'loss': 1.0055, 'learning_rate': 2.2222222222222223e-05, 'epoch': 0.29}                                             
+{'loss': 1.0081, 'learning_rate': 0.0, 'epoch': 0.32}                                                                
```

Comment:
The logging information `'learning_rate': 0.0` seems a little strange to me.

Reply:
```shell
GradScaler is enabled, but CUDA is not available.  Disabling.
  warnings.warn("torch.cuda.amp.GradScaler is enabled, but CUDA is not available.  Disabling.")
/opt/anaconda3/envs/yining-qlora/lib/python3.9/site-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(
{'loss': 1.7241, 'learning_rate': 0.0002, 'epoch': 0.03}                                                                    
{'loss': 1.317, 'learning_rate': 0.00017777777777777779, 'epoch': 0.06}                                                     
{'loss': 1.2302, 'learning_rate': 0.00015555555555555556, 'epoch': 0.1}                                                     
{'loss': 1.1478, 'learning_rate': 0.00013333333333333334, 'epoch': 0.13}                                                    
{'loss': 0.937, 'learning_rate': 0.00011111111111111112, 'epoch': 0.16}                                                     
{'loss': 0.9338, 'learning_rate': 8.888888888888889e-05, 'epoch': 0.19}                                                     
{'loss': 0.993, 'learning_rate': 6.666666666666667e-05, 'epoch': 0.22}                                                      
{'loss': 0.86, 'learning_rate': 4.4444444444444447e-05, 'epoch': 0.26}                                                      
{'loss': 1.0082, 'learning_rate': 2.2222222222222223e-05, 'epoch': 0.29}                                                    
{'loss': 1.0047, 'learning_rate': 0.0, 'epoch': 0.32}                                                                       
{'train_runtime': 245.3045, 'train_samples_per_second': 3.261, 'train_steps_per_second': 0.815, 'train_loss': 1.1155864334106445, 'epoch': 0.32}
```
Trained again and got the same result ? ( Didn‘t make change here